language/stdlib/modules/DesignatedDealer.move

```diff
@@ -121,8 +137,28 @@ module DesignatedDealer {
         add_tier<CoinType>(tc_account, dd_addr, TIER_3_DEFAULT * coin_scaling_factor);
     }
     spec fun add_currency {
-        /// >TODO: times out without any specification, perhaps it won't if fully specified.
-        pragma verify_duration_estimate = 100;
+        pragma opaque = true;
+        let dd_addr = Signer::spec_address_of(dd);
+        modifies global<TierInfo<CoinType>>(dd_addr);
+        ensures exists<TierInfo<CoinType>>(dd_addr);
+        ensures global<TierInfo<CoinType>>(dd_addr).window_start == LibraTimestamp::spec_now_microseconds();
```

Comment:
You can also do `ensures global<TierInfo<CoinType>>(add_addr) == TierInfo{window_start: .., window_inflow: ..}`. This ensures you have not forgotten any field, which is problematic with opaque.

Reply:
For this function, my specification for the tiers field is not exact; all I am saying is that its length is 4 without saying anything about the contents.  How do I use your suggestion in this situation?

Reply:
I think what Wolfgang is saying (correct me if I'm wrong please!) is you could do something like (not sure if this is the right syntax):
```
ensures global<TierInfo<CoinType>>(dd_addr) == TierInfo {
   window_start: LibraTimestamp::spec_now_microseconds(),
   window_inflow: 0,
   tiers: 4
};
```
That way it ensures that if a field is added to the `TierInfo` resource later that it will break the specification here so we know that it needs to be updated.

Reply:
tiers: 4 won't match the type of tiers, which is a vector.  I would have to write down in an expression exactly what the new value of tiers is.  My specification is deliberately partial.  It could be made more comprehensive if a client demands that.

Reply:
Although I have made the postcondition of add_tier and update_tier more precise, I have left the specification of add_currency unchanged for the reasons stated above.